### ISSUE TITLE
enhcompat.cc breaks using shared cudart

### ISSUE THREAD

**Issue by sclarkson** [NVIDIA/nccl#658]
Try compiling NCCL 2.12.7 with shared cudart, for example using the command below

`CUDARTLIB=cudart BUILDDIR=build-shared make`

After compilation completes, we can see that it is using the shared version of cudart

```
sclarkson@mcmurdo:~/nccl$ objdump -p build-shared/lib/libnccl.so.2.12.7 | grep NEEDED
  NEEDED               libcudart.so.11.0
  NEEDED               libpthread.so.0
  NEEDED               librt.so.1
  NEEDED               libdl.so.2
  NEEDED               libstdc++.so.6
  NEEDED               libm.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  NEEDED               ld-linux-x86-64.so.2
```

When running the NCCL tests with this compiled library,

```
sclarkson@mcmurdo:~/nccl-tests$ NCCL_DEBUG=WARN LD_LIBRARY_PATH=~/nccl/build-shared/lib/ ./build/all_gather_perf 
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 validation: 1 
#
# Using devices
#   Rank  0 Pid 165440 on    mcmurdo device  0 [0x81] NVIDIA GeForce GTX 1080 Ti
NCCL version 2.12.7+cuda11.4

mcmurdo:165440:165448 [0] init.cc:252 NCCL WARN Cuda failure 'CUDA driver is a stub library'
mcmurdo: Test NCCL failure common.cu:1100 'unhandled cuda error'
 .. mcmurdo pid 165440: Test failure common.cu:1007
```

Manually removing enhcompat.cc from the sources line in the Makefile resolves this issue, letting me use shared cudart as expected (and as worked prior  to 2.12.7).

Perhaps this compatibility shim should only be included in the static library?

**Comment by AddyLaddy** [comment]
Thanks for the report. This new enhanced compatibility code was added in NCCL 2.12.x 
@cliffwoolley may be able to shed a little more light on this.

**Comment by cliffwoolley** [comment]
The compatibility shim uses weak symbols and should work for the dynamic case as well.  It's actually important to have for dynamic cudart, just as with libnccl_static.a, as in either case you might have an older cudart at runtime than the one NCCL was built against. 

A few questions:
* Which compiler and which distro is this?
* Is there exactly one libcudart.so* version on your system, or are there several?  Which version(s) is/are they?  (I see mention there of cuda 11.4 but need additional info.)

**Comment by cliffwoolley** [comment]
(PS: out of curiosity, why dynamic cudart?  Static cudart is our standard procedure for NVIDIA's own dynamic library builds, FYI.  Not that I'm saying we shouldn't address whatever is going on here. )

**Comment by sclarkson** [comment]
I don't think you do need it for the dynamic libnccl.so, when you're statically compiling cudart in. The function calls in the code are guarded by a macro `CUDART_VERSION >= 11030` check. So, the only time the compat symbols would be used would be in the pathological case of compiling with >= 11.3 headers and linking to a < 11.3 library, no? And in doing that, you would simply create a non-functional shared library, as evidenced here.

Since the symbols are hidden, there is no opportunity for them to be overriden during runtime dynamic linking, as they are not weak in the emitted shared library.

```
sclarkson@mcmurdo:~/nccl$ nm build-shared/lib/libnccl.so.2.12.7 | grep cudaGetDriverEntryPoint
000000000005b950 t cudaGetDriverEntryPoint
```

To answer your other questions,

Ubuntu 20.04 with gcc-9
```
sclarkson@mcmurdo:~$ gcc --version
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Exactly one libcudart.so, version 11.4 for that test. Strace confirmed both the static and shared versions were opening the same libcuda.so, as well (version 510.54 for this test).

Also, I do realize this dynamic linking is of limited utility. But, I'm in an environment where everything that can be dynamically linked needs to be.

**Comment by benmenadue** [comment]
I just hit this compiling attempting to compile NCCL 2.12.12 on our cluster. We're similarly another site that uses dynamic libraries wherever possible. Static linking provides little benefit in most cases, and just makes providing common libraries for use by multiple users and across multiple applications and projects difficult.

The problem is that because `cudaGetDriverEntryPoint` is present in `libnccl.so` as a private hidden symbol, any reference to it from within `libnccl.so` is directly bound to it when linking the shared library; even if that name is provided by another library at runtime, it won't be used because it was resolved during the library creation stage.

And I don't think you can just export that symbol from your library because then you've got potentially got multiple definitions of the same symbol at runtime and you can end up with different libraries bound to different copies. That they are weak symbols doesn't change this: it will bind to the first it finds based on that library's search order regardless of whether it's strong or weak.

Perhaps a better solution is to put these compatibility stubs into a separate shared library, and make the calls to them from libnccl.so go through a wrapper that uses `dlsym(RTLD_DEFAULT)` to try to find the symbol based on what's already loaded; if a new enough `libcudart.so` is loaded, it will use those versions, otherwise it can `dlopen` the stub library and get them from there.

**Comment by cliffwoolley** [comment]
> Static linking provides little benefit in most cases,

For the CUDA Runtime, it actually provides substantial compatibility benefits.

But that's an aside to this issue, which I do intend to fix.  My current thinking is that the stubs can attempt to dlopen/dlsym libcudart.so on your behalf.  Should be easier than a second library, I think.

**Comment by cliffwoolley** [comment]
The other alternative is to just skip building the shim entirely when using dynamic CUDART -- i.e. the fix would be entirely handled at the Makefile level.  That has some appeal to it...

**Comment by benmenadue** [comment]
> For the CUDA Runtime, it actually provides substantial compatibility benefits.
Yes, I can see where it can make some things easier there because you don't need to worry about different versions -- you always use the one you were linked against. But then each shared library that depends on cudart will be using its own copy... how do you avoid incompatibilities between them? I guess as long as you don't have any global state or pass things from one copy to another it will be okay.

**Comment by benmenadue** [comment]
> skip building the shim entirely when using dynamic CUDART
Yeah. Then if you try to run with version of libcudart.so that's too old it will just fail on launch with an undefined symbol error. That's probably what I would prefer, actually, rather than have it fail later when trying to use it.

**Comment by cliffwoolley** [comment]
> But then each shared library that depends on cudart will be using its own copy... how do you avoid incompatibilities between them?

This is exactly what all NVIDIA-built libraries have done for years.  I've done quite a lot of mix-and-match CUDART version scenarios in applications I build; it works well.

> Yeah. Then if you try to run with version of libcudart.so that's too old it will just fail on launch with an undefined symbol error. 

Exactly.  I don't love it but maybe this is a K.I.S.S. scenario.  The reason we did enhcompat.cc in the first place was that if people used the NV-built static lib of NCCL (where they don't control which NVCC version we used) but their *own* static CUDART, and those versions were mismatched, then strange linker failures would result.  Here if the user builds their own NCCL and then mismatches themselves, hopefully the fallout would be more readily explained.
